dyninstAPI/src/linux.C

```diff
@@ -174,118 +180,167 @@ bool PCEventMuxer::useCallback(Dyninst::ProcControlAPI::EventType et)
-    if (dyn_path) { 
-       libPathStr = strdup(dyn_path);
-       libPath = strtok(libPathStr, ":");
-       while (libPath != NULL) {
-          libPaths.push_back(string(libPath));
-          libPath = strtok(NULL, ":");
-       }
-       free(libPathStr);
+    char* dyn_path = getenv("DYNINST_REWRITER_PATHS");
+    if(dyn_path)
+    {
+        for(const auto& itr : delimit(dyn_path, ":"))
+            _emplace_if_exists(itr);
     }
 
     // search paths from environment variables
-    char *ld_path = getenv("LD_LIBRARY_PATH");
-    if (ld_path) { 
-       libPathStr = strdup(ld_path);
-       libPath = strtok(libPathStr, ":");
-       while (libPath != NULL) {
-          libPaths.push_back(string(libPath));
-          libPath = strtok(NULL, ":");
-       }
-       free(libPathStr);
+    char* ld_path = getenv("LD_LIBRARY_PATH");
+    if(ld_path)
+    {
+        for(const auto& itr : delimit(ld_path, ":"))
+            _emplace_if_exists(itr);
     }
 
 #ifdef DYNINST_COMPILER_SEARCH_DIRS
     // search compiler-specific library paths
     // NB1: DYNINST_COMPILER_SEARCH_DIRS is defined at build time
     // NB2: This is explicitly done _after_ adding directories from
     //		LD_LIBRARY_PATH so that the user can override these paths.
-	{
-		#define xstr(s) str(s)
-		#define str(s) #s
-
-		libPathStr = strdup(xstr(DYNINST_COMPILER_SEARCH_DIRS));
-		libPath = strtok(libPathStr, ":");
-		while (libPath != NULL) {
-			libPaths.emplace_back(libPath);
-			libPath = strtok(NULL, ":");
-		}
-		free(libPathStr);
-
-		#undef str
-		#undef xstr
-	}
-#endif
-    //libPaths.push_back(string(getenv("PWD")));
-    for (unsigned int i = 0; i < libPaths.size(); i++) {
-        string str = libPaths[i] + "/" + filename;
-        if (stat(str.c_str(), &dummy) == 0) {
-            paths.push_back(str);
-        }
+    {
+#    define xstr(s) str(s)
+#    define str(s) #    s
+
+        for(const auto& itr : delimit(xstr(DYNINST_COMPILER_SEARCH_DIRS), ":"))
+            _emplace_if_exists(itr);
+
+#    undef str
+#    undef xstr
     }
+#endif
 
     // search ld.so.cache
     // apparently ubuntu doesn't like pclosing NULL, so a shared pointer custom
     // destructor is out. Ugh.
     FILE* ldconfig = popen("/sbin/ldconfig -p", "r");
-    if (ldconfig) {
-        if(!fgets(buffer, 512, ldconfig)) {	// ignore first line
-          return false;
-        }
-        while (fgets(buffer, 512, ldconfig) != NULL) {
-            pos = buffer;
-            while (*pos == ' ' || *pos == '\t') pos++;
-            key = pos;
-            while (*pos != ' ') pos++;
-            *pos = '\0';
-            while (*pos != '=' && *(pos + 1) != '>') pos++;
-            pos += 2;
-            while (*pos == ' ' || *pos == '\t') pos++;
-            val = pos;
-            while (*pos != '\n' && *pos != '\0') pos++;
-            *pos = '\0';
-            if (strcmp(key, filename.c_str()) == 0) {
-                paths.push_back(val);
+    if(ldconfig)
```

Comment:
@kupsch @bigtrak

This is just terrible, and I'm honestly surprised it took this long to get a nasty bug here. I vote we replace this and the manual directory list lookup below by just using

    if(void *h = dlopen(filename.c_str(), RTLD_LAZY)) {
      char origin[PATH_MAX];
      if(dlinfo(h, RTLD_DI_ORIGIN, name) == 0) {
        _emplace_if_exists(origin);  // check is useless, but uniform
      }
    }

The only caveat here is that

>If  the  calling object (i.e., the shared library or executable from which dlopen() is called) contains a DT_RPATH tag, and does not contain a DT_RUNPATH tag, then the directories listed in the DT_RPATH tag are searched.

If we want to work around this, we could just continue to do the manual search of LD_LIBRARY_PATH first to let users override.

Begrudgingly, the DYNINST_COMPILER_SEARCH_DIRS has to stay because ldconfig doesn't always have the compiler-specific search directories available. At least on my system, `dlopen` wouldn't find libgcc_s.so which lives in /usr/lib/gcc/x86_64-linux-gnu/12/libgcc_s.so.

With this, we'd have the heuristic:

1. Use the provided filename if file exists (via stat rules)
2. Search DYNINST_REWRITER_PATHS
3. Search LD_LIBRARY_PATH
4. Use ld lookup rules via `dlopen`.

---

`man dlopen` for reference

>o   (ELF  only)  If  the  calling object (i.e., the shared library or executable from which dlopen() is called) contains a DT_RPATH tag, and does not contain a DT_RUNPATH tag, then the directories listed in the DT_RPATH tag are searched.

>o   If, at the time that the program was started, the environment variable LD_LIBRARY_PATH was defined to contain a colon-separated list of directories, then these are searched.  (As a security measure, this variable is ignored for set-user-ID and set-group-ID programs.)

>o   (ELF only) If the calling object contains a DT_RUNPATH tag, then the directories listed in that tag are searched.

>o   The cache file /etc/ld.so.cache (maintained by ldconfig(8)) is checked to see whether it contains an entry for filename.

>o   The directories /lib and /usr/lib are searched (in that order).

Reply:
You want me to proceed with this? I'll probably first use with the RTLD_NOLOAD to make sure it's not already open bc I'm concerned whether or not there is a guarantee that loading the library will have any unintended side effects outside of just finding the library, e.g., cause Dyninst to start parsing that library.

Reply:
We had an internal discussion about this yesterday. The confounding factors are the RTLD_NOLOAD won't let you open a file unless it's already open. This means the only way to inspect RTLD_DI_ORIGIN is to actually load the library into Dyninst's address space. Aside from the security implications, it's just an inefficient way of doing it. We're pretty much stuck to parsing the output of ldconfig.

Reply:
What about NOLOAD on the library containing this function, grabbing that DI_ORIGIN and searching that? In many cases, that could give us the path we want. We could also NOLOAD dlopen the executable, get the link path, and search the dirs of the libs it is linked to

Reply:
i.e. something like here: https://github.com/AMDResearch/omnitrace/blob/9a665b5684c839130a2de3f5558a9f6c2c746e40/source/lib/binary/link_map.cpp#L98

Reply:
When using the RTLD_NOLOAD flag, a handle is not returned unless the object is already loaded.  Its purpose is to test if an object if loaded (or to change flags if loaded). Since there isn't a handle with the dlinfo data this doesn't help. I think ldconfig parsing is about as good as we can do.

Reply:
Yes, I fully understand what NOLOAD does. My point is that since this function is being called, unless the function is statically linked, the library is loaded so it _will_ return a handle. And the same goes for passing nullptr to dlopen (which returns the handle to the exe). As long as those are dynamic objects, you can get the handle via NOLOAD

Reply:
If this helps to clarify:

```cpp
auto* _handle = dlopen("libdyninstAPI.so", RTLD_LAZY|RTLD_NOLOAD");
// ... get origin of handle ...
auto _lib_directory = dirname(...);
```

In other words, the RT library you want is extremely likely to be installed in the same directory as the dyninstAPI library

Reply:
This function is supposed to resolve any library name, so we have to at least consider everything in ld.so.cache which is likely a larger set than the union of DYNINST_REWRITER_PATHS, LD_LIBRARY_PATH, DYNINST_COMPILER_SEARCH_DIRS, and $ORIGIN.

@kupsch @bigtrak 
This code is really just parsing the basename of the files reported from the cache. Why don't we just read the cache directly and report the basename or, if we want to implement a parser, from /etc/ld.so.conf?

Reply:
> This function is supposed to resolve any library name

Ok well, then how about adding a vector<string> hints parameter which gets searched after LD_LIBRARY_PATH for situations like finding the RT library. If you are concerned about the ABI, we can just implement one with the same signature that calls the new one with zero hints.

Reply:
>Ok well, then how about adding a vector hints parameter which gets searched after LD_LIBRARY_PATH for situations like finding the RT library.

What kinds of hints were you thinking? I think adding the "self" $ORIGIN would be good to do regardless. Refactoring this function to check for the file existence before proceeding on to the next lookup location would also be good. This would put off parsing ldconfig unless it's absolutely necessary and would make the $ORIGIN check an actual short circuit.

P.S. `dlopen(nullptr)` is guaranteed to give you the current binary whereas `dlopen("libdyninstAPI.so")` might not.